src/Voron/Impl/Journal/JournalWriter.cs

```diff
@@ -28,49 +28,96 @@ public unsafe class JournalWriter : IJournalWriter
             FileName = filename;
             _log = LoggingSource.Instance.GetLogger<JournalWriter>(options.BasePath.FullPath);
 
+            if (writeSynchronization == null)
+                writeSynchronization = new StorageEnvironmentSynchronization();
+
+            // The write synchronization mechanism will require a max size and to be aligned to 4kb.  
+            _writeSynchronization = writeSynchronization;
+            _maxNumberOf4KbBlocks = writeSynchronization.MaxWriteSize / (4L * Constants.Size.Kilobyte);
+
             var result = Pal.rvn_open_journal_for_writes(filename.FullPath, mode, size, options.SupportDurabilityFlags, out _writeHandle, out var actualSize, out var error);
             if (result != PalFlags.FailCodes.Success)
                 PalHelper.ThrowLastError(result, error, $"Attempted to open journal file - Path: {filename.FullPath} Size :{size}");
 
-            NumberOfAllocated4Kb = (int)(actualSize / (4 * Constants.Size.Kilobyte));
+            NumberOfAllocated4Kb = (int)(actualSize / (4L * Constants.Size.Kilobyte));
         }
 
-        public TimeSpan Write(long posBy4Kb, byte* p, int numberOf4Kb)
+        public TimeSpan Write(long posBy4Kb, byte* buffer, int numberOf4KbBlocks)
         {
             Debug.Assert(_options.IoMetrics != null);
 
             Stopwatch sp;
-
-            using (var metrics = _options.IoMetrics.MeterIoRate(FileName.FullPath, IoMetrics.MeterType.JournalWrite, numberOf4Kb * 4L * Constants.Size.Kilobyte))
+            using (var metrics = _options.IoMetrics.MeterIoRate(FileName.FullPath, IoMetrics.MeterType.JournalWrite, numberOf4KbBlocks * 4L * Constants.Size.Kilobyte))
             {
                 sp = Stopwatch.StartNew();
 
-                var result = Pal.rvn_write_journal(_writeHandle, p, numberOf4Kb * 4L * Constants.Size.Kilobyte, posBy4Kb * 4L * Constants.Size.Kilobyte, out var error);
-                if (result != PalFlags.FailCodes.Success)
-                    PalHelper.ThrowLastError(result, error, $"Attempted to write to journal file - Path: {FileName.FullPath} Size: {numberOf4Kb * 4L * Constants.Size.Kilobyte}, numberOf4Kb={numberOf4Kb}");
+                // We will the number of 4Kb aligned blocks in chunks of N blocks. 
+                long iterations = numberOf4KbBlocks / _maxNumberOf4KbBlocks;
+                long leftovers = numberOf4KbBlocks % _maxNumberOf4KbBlocks;
 
-                sp.Stop();
+                IoMeterBuffer.DurationMeasurement writeMetrics;
+
+                byte* currentBufferPtr = buffer;
+                for (long i = 0; i < iterations; i++ )
+                {
+                    writeMetrics = _options.IoMetrics.MeterIoRate(FileName.FullPath, IoMetrics.MeterType.JournalWait, _maxNumberOf4KbBlocks * 4L * Constants.Size.Kilobyte);
+                    using (_writeSynchronization.Enter())
+                    {
+                        writeMetrics.SetFileSize(_maxNumberOf4KbBlocks * (4L * Constants.Size.Kilobyte));
```

Comment:
is this correct? each iteration is writing less?

Reply:
No `_maxNumberOf4KbBlocks` comes from the constructor, it won't change. The first iteration will write always the max number of blocks possible through configuration. The only case that is not the case is when `leftovers` is non-zero. The only 2 conditions when it might happen is smaller than biggest block write or when we processed everything and we are left with a partial block at the end.